keanu-project/src/test/java/io/improbable/keanu/e2e/regression/RadonHeirarchicalRegression.java

```diff
@@ -0,0 +1,179 @@
+package io.improbable.keanu.e2e.regression;
+
+import io.improbable.keanu.DeterministicRule;
+import io.improbable.keanu.algorithms.variational.optimizer.gradient.GradientOptimizer;
+import io.improbable.keanu.model.regression.RegressionModel;
+import io.improbable.keanu.model.regression.RegressionRegularization;
+import io.improbable.keanu.network.BayesianNetwork;
+import io.improbable.keanu.tensor.dbl.DoubleTensor;
+import io.improbable.keanu.util.csv.ReadCsv;
+import io.improbable.keanu.vertices.VertexLabel;
+import io.improbable.keanu.vertices.dbl.DoubleVertex;
+import io.improbable.keanu.vertices.dbl.probabilistic.GaussianVertex;
+import io.improbable.keanu.vertices.dbl.probabilistic.HalfGaussianVertex;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+public class RadonHeirarchicalRegression {
+
+    @Rule
+    public DeterministicRule rule = new DeterministicRule();
+
+    private List<Data> radonData;
+
+    @Before
+    public void readRadonCSV() {
+        radonData = ReadCsv.fromResources("data/datasets/radon/radon.csv")
+            .asRowsDefinedBy(Data.class)
+            .load(true);
+    }
+
+    @Test
+    public void canPerformSimpleLinearRegression() {
+        RegressionModel model = linearRegression(radonData);
+        Assert.assertTrue(model.getWeight(0) > -0.7 && model.getWeight(0) < -0.4);
+        Assert.assertTrue(model.getIntercept() > 1.2 && model.getIntercept() < 1.5);
+    }
+
+    @Test
+    public void canPerformRegressionWithOneHeirarchy() {
+        buildHeirarchicalNetwork(radonData, 1);
+    }
+
+    @Test
+    public void canPerformRegressionWithTwoHeirarchies() {
+        buildHeirarchicalNetwork(radonData, 2);
+    }
+
+    @Test
+    public void canPerformRegressionWithFourHeirarchies() {
+        buildHeirarchicalNetwork(radonData, 4);
+    }
+
+    @Test
+    public void canPerformRegressionWithTenHeirarchies() {
+        buildHeirarchicalNetwork(radonData, 10);
+    }
+
+    private RegressionModel linearRegression(List<Data> data) {
+        double[] radon = data.stream().mapToDouble(k -> k.log_radon).toArray();
+        double[] floor = data.stream().mapToDouble(k -> k.floor).toArray();
+        DoubleTensor y = DoubleTensor.create(radon, 1, radon.length);
+        DoubleTensor x = DoubleTensor.create(floor, 1, floor.length);
+
+        RegressionModel model = RegressionModel.
+            withTrainingData(x, y).
+            withRegularization(RegressionRegularization.RIDGE).
+            withPriorOnWeights(0., 5.).
+            withPriorOnIntercept(0., 5.).
+            build();
+
+        model.observe();
+        model.fit();
+
+        return model;
+    }
+
+    private void buildHeirarchicalNetwork(List<Data> radonData, int numberOfModels) {
```

Comment:
This builds the network and runs an optimiser on it. The name should reflect that.

Reply:
It also does all the verification.